### PR TITLE
Parsed vector aux kernel

### PR DESCRIPTION
**Pull request by GiudGiud**
Refs #24380 

This allows regular, vector aux and xyzt to be coupled into a vector aux variable

### CODE REVIEW

File: framework/include/interfaces/Coupleable.h [hunk 1]
```diff
@@ -1553,6 +1560,14 @@ class Coupleable
    */
   const ArrayMooseVariable * getArrayVar(const std::string & var_name, unsigned int comp) const;
 
+  /**
+   * Extract pointer to a base class for a coupled variable of any type
+   * @param var_name Name of parameter desired
+   * @param comp Component number of multiple coupled variables
+   * @return Pointer to the desired variable
+   */
+  const MooseVariableFieldBase * getAnyVar(const std::string & var_name, unsigned int comp) const;
```
Comment (lindsayad) [inline framework/include/interfaces/Coupleable.h:1569]: We already have `getFieldVar`

File: framework/src/auxkernels/ParsedVectorAux.C [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    _use_xyzt(getParam<bool>("use_xyzt"))
+{
+  _func_F.resize(LIBMESH_DIM);
+  _function.resize(LIBMESH_DIM);
+
+  if (getParam<std::vector<std::vector<std::string>>>("constant_names").size() !=
+      getParam<std::vector<std::vector<std::string>>>("constant_expressions").size())
+    paramError("constant_names", "Must be same length as constant_expressions");
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
```
Comment (lindsayad) [inline framework/src/auxkernels/ParsedVectorAux.C:73]:
```suggestion
  for (const auto i : make_range(Moose::dim))
```

File: framework/src/auxkernels/ParsedVectorAux.C [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  {
+    // build variables argument
+    std::string variables;
+
+    // coupled regular field variables
+    for (std::size_t i = 0; i < _nargs; ++i)
+      variables += (i == 0 ? "" : ",") + getFieldVar("coupled_variables", i)->name();
+
+    // coupled vector field variables
+    for (std::size_t i = 0; i < _n_vector_args; ++i)
```
Comment (lindsayad) [inline framework/src/auxkernels/ParsedVectorAux.C:83]:
```suggestion
    for (const auto i : make_range(_nargs))
      variables += (i == 0 ? "" : ",") + getFieldVar("coupled_variables", i)->name();

    // coupled vector field variables
    for (const auto i : make_range(_n_vector_args))
```

File: framework/src/auxkernels/ParsedVectorAux.C [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  {
+    for (std::size_t j = 0; j < _nargs; ++j)
+      _func_params[j] = (*_args[j])[_qp];
+
+    for (std::size_t j = 0; j < _n_vector_args; ++j)
+      _func_params[_nargs + j] = (*_vector_args[j])[_qp](i);
+
+    if (_use_xyzt)
+    {
+      for (std::size_t j = 0; j < LIBMESH_DIM; ++j)
```
Comment (lindsayad) [inline framework/src/auxkernels/ParsedVectorAux.C:155]:
```suggestion
  for (const auto i : make_range(Moose::dim))
  {
    for (const auto j : make_range(_nargs))
      _func_params[j] = (*_args[j])[_qp];

    for (const auto j : make_range(_n_vector_args))
      _func_params[_nargs + j] = (*_vector_args[j])[_qp](i);

    if (_use_xyzt)
    {
      for (const auto j : make_range(Moose::dim))
```

File: framework/src/interfaces/Coupleable.C [hunk 5]
```diff
@@ -325,6 +325,37 @@ Coupleable::getArrayVar(const std::string & var_name, unsigned int comp) const
...
+    }
+  }
+
+  auto coupled_vars_it = _coupled_vars.find(name_to_use);
+
+  mooseAssert(coupled_vars_it != _coupled_vars.end(),
+              "Trying to get a coupled var " << name_to_use << " that doesn't exist");
+
+  return coupled_vars_it->second[comp];
+}
```
Comment (lindsayad) [inline framework/src/interfaces/Coupleable.C:357]: Delete or consolidate with `getFieldVar`

File: framework/src/interfaces/Coupleable.C [hunk 6]
```diff
@@ -2310,8 +2341,8 @@ Coupleable::coupledIndices(const std::string & var_name) const
 VariableName
 Coupleable::coupledName(const std::string & var_name, unsigned int comp) const
 {
-  if (getVar(var_name, comp))
-    return getVar(var_name, comp)->name();
+  if (getAnyVar(var_name, comp))
+    return getAnyVar(var_name, comp)->name();
```
Comment (lindsayad) [inline framework/src/interfaces/Coupleable.C:2345]:
```suggestion
  if (getFieldVar(var_name, comp))
    return getFieldVar(var_name, comp)->name();
```

File: test/tests/auxkernels/parsed_vector_aux/tests [hunk 7]
```diff
@@ -0,0 +1,12 @@
+[Tests]
+  design = 'source/auxkernels/ParsedAux.md'
+  [test]
+    type = 'Exodiff'
+    input = 'parsed_aux_test.i'
+    exodiff = 'out.e'
+    issues = '#4807'
+    requirement = "The system shall be able to compute vector values for vector from a functional "
+                  "expression."
```
Comment (lindsayad) [inline test/tests/auxkernels/parsed_vector_aux/tests:9]: "compute vector values for vector" is unclear to me

**Review by lindsayad — CHANGES_REQUESTED** [review]
@dschwen please feel free to take a look at the parsed function stuff in this PR

File: framework/src/auxkernels/ParsedVectorAux.C [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+      "FunctionExpression",
+      "Parsed function expression to compute the x component");
+  params.addCustomTypeParam<std::string>("expression_y",
+                                         "0",
+                                         "FunctionExpression",
+                                         "Parsed function expression to compute the y component");
+  params.addCustomTypeParam<std::string>("expression_z",
+                                         "0",
+                                         "FunctionExpression",
+                                         "Parsed function expression to compute the z component");
```
Comment (dschwen) [inline framework/src/auxkernels/ParsedVectorAux.C:34]:
```suggestion
  if constexpr (LIBMESH_DIM >= 3) 
    params.addCustomTypeParam<std::string>("expression_z",
                                           "0",
                                           "FunctionExpression",
                                           "Parsed function expression to compute the z component");
```

File: test/tests/auxkernels/parsed_vector_aux/tests (hunk 7 above)
Comment (GiudGiud) [reply test/tests/auxkernels/parsed_vector_aux/tests]: reworded

File: framework/src/interfaces/Coupleable.C (hunk 5 above)
Comment (GiudGiud) [reply framework/src/interfaces/Coupleable.C]: getFieldVar is just better, deleted getAnyVar

File: framework/include/interfaces/Coupleable.h (hunk 1 above)
Comment (GiudGiud) [reply framework/include/interfaces/Coupleable.h]: removed getAnyVar

**Review by lindsayad — APPROVED** [review]

**Comment by GiudGiud** [comment]
Thanks for the review!

**Pull request merged**